### PR TITLE
ISPN-12931 Server configuration overlay

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12931

### CODE REVIEW

**Review by ryanemerson — COMMENTED** [review]
I really like the simplicity of this :slightly_smiling_face:

I think we will have an issue if the currently processed configuration layer does not create a "valid" config. This is due to some of the semantic checks that exist in the parser. Ensuring that semantic validation occurs entirely at configuration build time would allow this. I don't think this is part of this PR though, just a nice to have in the future for more flexibility.

From a dev/support perspective it would be great if the "final" configuration was also output to the filesystem (xml is fine) so that it can easily be debugged.

**Comment by ryanemerson** [comment]
> I think we will have an issue if the currently processed configuration layer does not create a "valid" config. This is due to some of the semantic checks that exist in the parser. Ensuring that semantic validation occurs entirely at configuration build time would allow this. I don't think this is part of this PR though, just a nice to have in the future for more flexibility.

To give a more concrete example, if I want to add an additional `socket-binding` via an overlay it's necessary for me to provide the mandatory `socket-bindings` attributes in the overlay.


```
    socketBindings:
#      defaultInterface: public
#      portOffset: 0
      bindings:
        - socketBinding:
            name: admin
            port: 11223
```

If the `defaultInterface` and `port-offset` attributes are missing, then we get the following:

```
2021-04-12 15:13:05,648 FATAL (main) [org.infinispan.SERVER] ISPN080028: Infinispan Server failed to start org.infinispan.commons.configuration.io.ConfigurationReaderException: Missing required attribute(s): default-interface[45,1]
```

**Comment by tristantarrant** [comment]
@ryanemerson moving the validation from the parser to the config objects is definitely material for a separate

**Comment by ryanemerson** [comment]
Agreed. 

`ServerConfigurationParserTest.xml` failure is related.

> From a dev/support perspective it would be great if the "final" configuration was also output to the filesystem (xml is fine) so that it can easily be debugged.

How difficult would this be to implement? It really would help during the initial dev and debugging the operator when users are providing custom configurations. \\cc @rigazilla

**Comment by ryanemerson** [comment]
Thanks @tristantarrant. The DEBUG log is sufficient for now.

File: commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java [hunk 1]
```diff
@@ -44,22 +44,8 @@ public YamlConfigurationReader(Reader reader, ConfigurationResourceResolver reso
       this.reader = reader instanceof BufferedReader ? (BufferedReader) reader : new BufferedReader(reader);
       namespaces.put("", ""); // Default namespace
       loadTree();
-      printTree(lines);
    }
 
-   private void printTree(Node node) {
```
[the remark this answers is not in the document]
Comment (danberindei) [reply commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java:50]: I hope we don't need it :)

**Review by danberindei — COMMENTED** [review]
I don't really like the idea of free-form updates to the same cache configuration from multiple full configuration files, I feel like this could get hard to understand. Although I guess the fact that an overlay could define a wildcard template that applies to the cache or overrides another template that was applying to the cache makes the understanding of overlay parsing difficult enough anyway...

I guess it's all good because we have a way for the user to get the "effective" global/cache configuration and see what was actually applied.

Update: I was looking only at the diff and I missed the discussion here. Indeed, a debug log looks like the best option for servers that are not able to start because of configuration problems.

**Pull request merged**